dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java

```diff
@@ -54,641 +46,238 @@
-        }
-
-        if (includedSystemHostOnLuceneQuery && !Host.SYSTEM_HOST.equals(siteId) &&
-                null == this.vanityUrlServices.getCachedVanityUrlList(Host.SYSTEM_HOST, languageId)) {
-
-            this.vanityUrlServices.setCachedVanityUrlList(Host.SYSTEM_HOST, languageId, Collections.EMPTY_LIST);
-        }
-    } // checkSiteLanguageVanities.
-
-    private void setEmptyCaches(final String siteId,
-                                final Long languageId,
-                                final Boolean includedSystemHostOnLuceneQuery) {
-
-        if (null != siteId && null != languageId) {
-            this.vanityUrlServices.setCachedVanityUrlList(
-                    siteId, languageId,
-                    Collections.EMPTY_LIST);
-
-            if (includedSystemHostOnLuceneQuery && !Host.SYSTEM_HOST.equals(siteId)) {
-                this.vanityUrlServices.setCachedVanityUrlList(
-                        Host.SYSTEM_HOST, languageId,
-                        Collections.EMPTY_LIST);
-            }
-        }
-    } // setEmptyCaches.
-
-    private void addSecondaryVanityURLCacheCollection(final List<VanityUrl> vanityUrls) {
-
-        final Map<SiteLanguageKey, ImmutableList.Builder<CachedVanityUrl>> vanityPerSiteLanguageMap
-                = map();
-        SiteLanguageKey key;
-
-        for (final VanityUrl vanityUrl : vanityUrls) {
-
-            key = new SiteLanguageKey(vanityUrl.getSite(), vanityUrl.getLanguageId());
-            if (!vanityPerSiteLanguageMap.containsKey(key)) {
-
-                vanityPerSiteLanguageMap.put(key, new ImmutableList.Builder());
-            }
-
-            vanityPerSiteLanguageMap.get(key).add(new CachedVanityUrl(vanityUrl));
-        }
-
-        // when the site + lang list is gonna be override, the main cache items should be clean too
-        vanityPerSiteLanguageMap.keySet().forEach(this::cleanCurrentVanityUrlsPerSite);
-
-        vanityPerSiteLanguageMap.forEach( (k, vanityUrlBuilder) ->
-                    this.vanityUrlServices.setCachedVanityUrlList (k.hostId(), k.languageId(), vanityUrlBuilder.build()));
-    } // addSecondaryVanityURLCacheCollection.
-
-    private void cleanCurrentVanityUrlsPerSite(final SiteLanguageKey siteLanguageKey) {
-
-        final List<CachedVanityUrl> cachedVanityUrls = this.vanityUrlServices.getCachedVanityUrlList
-                (siteLanguageKey.hostId(), siteLanguageKey.languageId());
-
-        if (null != cachedVanityUrls) {
-
-            for (final CachedVanityUrl cachedVanityUrl : cachedVanityUrls) {
-
-                CacheLocator.getVanityURLCache().remove(new CacheVanityKey(
-                        cachedVanityUrl.getSiteId(),
-                        cachedVanityUrl.getLanguageId(),
-                        cachedVanityUrl.getUrl()
-                ).toString());
-            }
-        }
+    // 404 short circuit
+    Optional<CachedVanityUrl> shortCircuit = cache.getDirectMapping(url, host, lang);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Local variable 'shortCircuit' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5434315)